db/write_batch.cc

```diff
@@ -1190,6 +1190,21 @@ Status WriteBatch::PopSavePoint() {
   return Status::OK();
 }
 
+Status WriteBatch::AssignTimestamp(
+    const Slice& ts, std::function<Status(uint32_t, size_t&)> checker) {
+  TimestampAssigner<decltype(checker)> ts_assigner(prot_info_.get(), checker,
```

Comment:
Here we're actually using both static and dynamic polymorphism at the same time when either one would suffice by itself. I think we could either

1) Use static polymorphism (templates) and move all template definitions etc. to a header. This doesn't have to be `write_batch.h` itself; it could "semi-public" / "impl" header that would get `#include`d by `write_batch.h`. (We could also put `TimestampAssigner` and friends in an "impl" or "detail" namespace to make it clear they're not for direct application use.)
2) Use dynamic polymorphism (e.g. `std::function`) but then we don't need any of the template magic we added. This has a runtime cost though.

Reply:
> "semi-public" / "impl" header

I searched all the existing public headers, but none of them does this at the moment. I am concerned about its consequences, e.g. binary size, maintainability, etc. In this case, we need to consider including stuff from `db/kv_checksum.h`, `db/dbformat.h`, `util/hash.h`, etc., which are non-public. The size of the change is bigger than what I desire.

The trick of making `TimestampAssignerBase` a template saves the virtual call incurred by `TimestampAssignerBase::AssignTimestamp()`, which is orthogonal to the use of `std::function`'s type erasure cost. Maybe I am missing something?

Reply:
> The trick of making `TimestampAssignerBase` a template saves the virtual call incurred by `TimestampAssignerBase::AssignTimestamp()`, which is orthogonal to the use of `std::function`'s type erasure cost. Maybe I am missing something?

I'm referring to the `Checker` template parameter, which seems unnecessary if we're only instantiating the templates with a single `Checker` type: `std::function<Status(uint32_t, size_t&)>`. We can still use CRTP to avoid having to make `AssignTimestampImpl` a virtual.

Reply:
Thanks for the clarification. Will remove that for now.

Reply:
Yeah, sorry, saying that we don't need *any* of the template magic was not accurate :)